### PR TITLE
feat: notify on virtual level ups

### LINKED ISSUE #31: Add Virtual Level up notifications
Send virtual level notifications based on current XP
[https://oldschool.runescape.wiki/w/Experience#Virtual_levels](url)

### PR DESCRIPTION
**Pull request by iProdigy**
Closes #31 (related upstream issue: https://github.com/MidgetJake/UniversalDiscordNotifier/issues/17)

### CODE REVIEW

File: src/main/java/dinkplugin/LevelNotifier.java [hunk 1]
```diff
@@ -79,15 +82,36 @@ public void attemptNotify() {
...
     }
 
-    public void handleLevelUp(String skill, int level) {
+    public void handleLevelUp(String skill, int level, int xp) {
         if (plugin.isSpeedrunWorld()) return;
-        if (plugin.config.notifyLevel() && checkLevelInterval(level) && currentLevels.get(skill) != null) {
-            if (level == currentLevels.get(skill)) {
-                return;
+
+        int virtualLevel = level < 99 ? level : getLevelForExperience(xp);
```
Comment (iProdigy) [inline src/main/java/dinkplugin/LevelNotifier.java:88]: It's technically possible to add a toggleable setting for this but: on toggle, one would need to call `#reset` or we'd need to keep a map of xp per skill to be able to switch between virtual and non-virtual levels

let me know if this should be implemented in this PR

File: src/main/java/dinkplugin/LevelNotifier.java [hunk 2]
```diff
@@ -79,15 +82,36 @@ public void attemptNotify() {
...
+                sendMessage = true;
             }
-            levelledSkills.add(skill);
-            sendMessage = true;
         }
-        currentLevels.put(skill, level);
+    }
+
+    private static Integer getLevelForExperience(int xp) {
+        return xp > 0 ? LEVEL_BY_EXP_THRESHOLD.floorEntry(xp).getValue() : 1;
```
Comment (iProdigy) [inline src/main/java/dinkplugin/LevelNotifier.java:99]: $O(log(n))$ lookup; the alternative is hard-coding the [thresholds](https://oldschool.runescape.wiki/w/Experience#Virtual_levels) in a bunch of `if` statements

File: src/main/java/dinkplugin/LevelNotifier.java (hunk 1 above)
Comment (pajlada) [reply src/main/java/dinkplugin/LevelNotifier.java:88]: I imagine that if you have levelups enabled, you wouldn't mind sending a notification on virtual levelups

File: src/main/java/dinkplugin/LevelNotifier.java [hunk 3]
```diff
@@ -79,15 +82,36 @@ public void attemptNotify() {
...
+    }
+
+    static {
+        final int minLevel = 1, maxLevel = 126;
+
+        LEVEL_BY_EXP_THRESHOLD = new TreeMap<>();
```
Comment (pajlada) [inline src/main/java/dinkplugin/LevelNotifier.java:105]: In the RuneLite API, there's a [getXpForLevel](https://static.runelite.net/api/runelite-api/net/runelite/api/Experience.html#getXpForLevel(int)) function which functions very similarly, we should be able to use that instead

Comment (iProdigy) [reply src/main/java/dinkplugin/LevelNotifier.java]: Good point - simplified the code with `getLevelForXp`

**Comment by Felanbird** [comment]
I'll test this in 300k defence xp

**Review by Felanbird — APPROVED** [review]
image

**Review by pajlada — APPROVED** [review]

**Pull request merged**